jina/peapods/runtime.py

```diff
@@ -0,0 +1,180 @@
+    _dct = {}
+
+    def __new__(cls, name, bases, dct):
+        _cls = super().__new__(cls, name, bases, dct)
+        RuntimeMeta._dct.update({name: {'cls': cls,
+                                        'name': name,
+                                        'bases': bases,
+                                        'dct': dct}})
+        return _cls
+
+    def __call__(cls, *args, **kwargs) -> 'RuntimeMeta':
+        # switch to the new backend
+        _cls = {
+            'thread': threading.Thread,
+            'process': multiprocessing.Process,
+        }.get(getattr(args[0], 'runtime', 'thread'))
+
+        # rebuild the class according to mro
+        for c in cls.mro()[-2::-1]:
+            arg_cls = RuntimeMeta._dct[c.__name__]['cls']
+            arg_name = RuntimeMeta._dct[c.__name__]['name']
+            arg_dct = RuntimeMeta._dct[c.__name__]['dct']
+            _cls = super().__new__(arg_cls, arg_name, (_cls,), arg_dct)
+
+        return type.__call__(_cls, *args, **kwargs)
+
+
+class RunTime(metaclass=RuntimeMeta):
+    def __init__(self, args: Union['argparse.Namespace', Dict]):
+        self.args = args
+        self.name = self.__class__.__name__
+
+        self.is_ready_event = _get_event(self)
+        self.is_shutdown = _get_event(self)
+        self.ready_or_shutdown = _make_or_event(self, self.is_ready_event, self.is_shutdown)
+        self.is_shutdown.clear()
+
+        if isinstance(self.args, argparse.Namespace):
+            if self.args.name:
+                self.name = f'support-{self.args.name}'
+            elif self.args.role == PeaRoleType.PARALLEL:
+                self.name = f'support-{self.name}-{self.args.pea_id}'
+
+            self.ctrl_addr, self.ctrl_with_ipc = Zmqlet.get_ctrl_address(self.args.host, self.args.port_ctrl,
+                                                                         self.args.ctrl_with_ipc)
+            self.logger = JinaLogger(self.name,
+                                     log_id=self.args.log_id,
+                                     log_config=self.args.log_config)
+        else:
+            self.logger = JinaLogger(self.name)
+
+        self.pea = Pea(args, allow_remote=True)
+
+    def set_ready(self, *args, **kwargs):
+        """Set the status of the pea to ready """
+        self.is_ready_event.set()
+        self.logger.success(__ready_msg__)
+
+    def unset_ready(self, *args, **kwargs):
+        """Set the status of the pea to shutdown """
+        self.is_ready_event.clear()
+        self.logger.success(__stop_msg__)
+
+    def run(self):
+        """Start the request loop of this BasePea. It will listen to the network protobuf message via ZeroMQ. """
+        try:
+            with Pea(self.args) as pea:
+                # TODO: set_ready in different coroutine checking status as it is done for `ContainerPea` (here zmq
+                #  loop has not started)
```

Comment:
We should also have an option to configure an event loop (default asyncio or uvloop). This will help in having custom loops in each process.

Reply:
I think this does not belong in this PR right?